JSON Logging Take Two

ZephyrTFA opened the pull request:
## About The Pull Request

Converts all logging, excluding perf and investigate, to json.
I focused on making the system as easy to use and as easy to add new categories as possible.

Due to issues related to logging to world at global creation logger is now a byond real, which is created directly before Master

Log categories support versioning, secret flagging, and sub-category filtering. Although all of this is entirely optional for coders.
If you ever want to add a new category and use it, all you need to do is make the barebones category datum and the define.
I've kept existing procs such as log_game, and simply turned them into a wrapper for Logger.Log(xxx, ...)
## Why It's Good For The Game

Makes processing and filtering logs much easier in the future, while only minimally downgrading log crawling experience.
I am also working on a log viewer frontend for admin usage however that will take a little bit longer to finish up.
Also makes special logging and data tracking much easier thanks to a data list processing implementation and handling
## Changelog
:cl:
server: All logs are now formatted in json, excluding perf and investigations
/:cl:

ZephyrTFA committed:
add sub category handling and schema tracking
also add secret flagging

ZephyrTFA committed:
game category

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into jlog

ZephyrTFA committed:
finish up game and admin logging

ZephyrTFA committed:
protect logger from admin manipulation

ZephyrTFA committed:
more

ZephyrTFA committed:
even more logs

ZephyrTFA committed:
this is a typecache not a normal list

ZephyrTFA committed:
forgot to set master category on these

ZephyrTFA committed:
the rest

ZephyrTFA committed:
recursive jsonification

ZephyrTFA committed:
some serialize impls

ZephyrTFA committed:
more serialization logic

Mothblocks commented on code/datums/memory/_memory.dm:
```diff
@@ -75,6 +75,23 @@
 	memorizer_mind = null
 	return ..()
 
+/datum/memory/serialize_list(list/options)
```
When did we commit this API? We need semver versioning on this

Mothblocks commented on code/modules/logging/log_holder.dm:
```diff
@@ -1,5 +1,4 @@
-GLOBAL_DATUM_INIT(logger, /datum/log_holder, new)
-GLOBAL_PROTECT(logger)
+GLOBAL_REAL(Logger, /datum/log_holder)
```
1. Why is this a GLOBAL_REAL?
2. It should've stayed `logger`

Mothblocks replied:
Ugh the convenience here might be important enough, I want a global proc for this instead but I can't even think of a name that isn't `log` (and `Log` annoys me)

ZephyrTFA replied:
> Due to issues related to logging to world at global creation Logger is a global real, which is setup in the same manner Config and GLOB is, by having Master do it in Initializations

Mothblocks commented on code/modules/logging/log_holder.dm:
```diff
@@ -14,51 +13,136 @@ GLOBAL_PROTECT(logger)
+	)
+	rustg_file_write("[json_encode(category_header)]\n", category_instance.get_output_file(null))
+
 /// This is Log because log is a byond internal proc
```
This shouldn't be a doc comment btw, this is an implementation detail.

This needed an actual doc comment

ZephyrTFA committed:
change to logger, add code doc, save impl information and check for invalid category

MrStonedOne commented on code/controllers/master.dm:
```diff
@@ -84,6 +84,8 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 	if(!config)
 		config = new
+	if(!Logger)
+		Logger = new
```
Nope. not the right spot for this. When i did my first refactor New and Initialize on master had hundreds of snowflake lines and I am very protective of people dumping world start lines in New(). config is only allowed because it requires the MC existing to work. Logger must not require the MC to exist, or Master to non-null (or hell, it must not require config to exist or be non-null. log types without a config type must work in those situations)

Please go to line 14 in this same file and see how the master and debugger vars are setup, and do the same with logger.

ZephyrTFA committed:
add some basic semver logic for serialization

ZephyrTFA committed:
whoops

MrStonedOne commented on code/modules/logging/log_holder.dm:
```diff
@@ -14,51 +11,143 @@ GLOBAL_PROTECT(logger)
 /datum/log_holder/proc/Log(category, message, list/data)
-	if(!initialized || shutdown)
-		CRASH("Attempted to perform logging before initializion or after shutdown!")
+	// This is Log because log is a byond internal proc
+	if(shutdown)
+		CRASH("Attempted to perform logging after shutdown!")
+
+	if(!istext(message))
+		CRASH("Attempted to log a non-text message! [message]")
+
+	if(!category)
+		CRASH("Attempted to log to a null category! [message]")
+
+	if(data && !islist(data))
+		CRASH("Attempted to log a non-list data! [data]")
```
None of these are allowed.

logging has to be very liberal in what it accepts and must try as hard as it can to log.

also you can't have logging messages going to the runtime log given they may contain secret logs.

for after shutdown, it still has to log to disk immediately.

for non-text messages, it should stringify the message and log. (with a warning)

for non-category messages, it should default to the LOG_CATEGORY_NOT_FOUND category but then prepend the category to the start message (incase the category includes the message)

for non-list data, it should either stringify the data and append it to the end the message or turn data into `list("data" = data)`

MrsTonedOne reviewed (changes requested):
Another thing i noticed:

the json logging system does not include timestamps with the same verbosity as the rustg txt log timestamps.

This must be fixed. Logs must be mergeable and sortable without requiring the use of a stable sorting engine. 

I had milliseconds added to rustg's logging timestamp for a reason.

also you can't use unixtime as a number in byond. the resolution goes byond floats whole-number-precision and it ends up rounding to whole ass minutes.

edit: The current unixtime number for right now, when put in a float32 as a whole number of seconds, has 7 bits of precision loss, or 128 seconds.

optimumtact commented:
human readable should be an ISO timestamp debate me

MrStonedOne commented:
> human readable should be an ISO timestamp debate me
> 
> also rather than get strings from rust-g you could just have rust-g write the logs and decorate the extra info itself.

these also are stored in lists and the current round's logs are browsable to admins to some extent currently so the game code does need to know a real timestamp it can show the admins, that ideally would match the log files incase they try to convert from the ingame tool to the same log they are viewing outside of the game, as they should be able to control+f for a timestamp from one to find the same line in the log. 

(btw, storing these during the round is highly likely to be a shitty idea in this expanded version of json logs. i haven't done the math on it but i would not be shocked if this lead to some insanely high memory usage issues in production. given what i know about current round log lengths)

Mothblocks commented on code/datums/martial/_martial.dm:
```diff
@@ -16,6 +16,13 @@
 	var/pacifist_style = FALSE
 
+/datum/martial_art/serialize_list(list/options)
+	. = list()
```
Should be doing `return .` with all of these since theyre new APIs

Mothblocks replied:
You resolved this but didn't do it

Mothblocks commented on code/datums/memory/_memory.dm:
```diff
@@ -75,6 +75,25 @@
+		.["antagonist_name"] = antagonist_name
+	if(where)
+		.["where"] = where
+	options[SCHEMA_VERSION] = "1.0"
```
This isn't semver. Semver is 1.0.0.

Mothblocks replied:
The patch is important in case we need to do something like fix a property coming out as null or something.

Mothblocks replied:
Make sure to assert that SCHEMA_VERSION is valid semver (in our case, just major.minor.patch)

Mothblocks commented on code/modules/logging/log_holder.dm:
```diff
@@ -14,51 +11,143 @@ GLOBAL_PROTECT(logger)
-	if(!initialized || shutdown)
-		CRASH("Attempted to perform logging before initializion or after shutdown!")
+	// This is Log because log is a byond internal proc
+	if(shutdown)
+		CRASH("Attempted to perform logging after shutdown!")
```
Nit (especially since MSO wants you to remove these anyway): You can use ASSERT for these now and it's a lot nicer

`ASSERT(condition, "message goes here")`

Mothblocks commented on code/modules/photography/_pictures.dm:
```diff
@@ -68,8 +68,11 @@
+	options[SCHEMA_VERSION] = "1.0"
 
 /datum/picture/deserialize_list(list/input, list/options)
+	if(options[SCHEMA_VERSION] != "1.0")
```
This serialize_list and deserialize_list stuff was added by JSON logging, right? We're not about to make it so nobodys persisted photo albums can't load?

Mothblocks replied:
You didn't answer this. What is going to happen to people's pictures?

Mothblocks commented on code/modules/logging/log_holder.dm:
```diff
@@ -67,4 +156,37 @@ GLOBAL_PROTECT(logger)
+				SCHEMA_VERSION = LOG_CATEGORY_SCHEMA_VERSION_NOT_SET,
+			)
+			var/list/serialization_data = data.serialize_list(options_list)
+			if(!length(serialization_data)) // serialize_list wasn't implemented, and errored
```
Can we stack_trace here just in case we're dealing with a shitty `return null` or something else stupid?

Mothblocks commented on code/modules/logging/log_holder.dm:
```diff
@@ -14,51 +11,143 @@ GLOBAL_PROTECT(logger)
+/datum/log_holder/proc/assemble_log_category_tree()
+	var/static/list/category_tree
+	if(category_tree)
+		return category_tree.Copy()
```
Why are we copying here? If it is needed, it should be done by consumers and not this API

ZephyrTFA committed:
few changeups

ZephyrTFA committed:
semver validation and dont copy list

ZephyrTFA committed:
verify semver on init

ZephyrTFA committed:
some redos

ZephyrTFA committed:
forgor

ZephyrTFA committed:
now are you happy byond

ZephyrTFA committed:
yeah that's the issue

ZephyrTFA committed:
ya

ZephyrTFA committed:
add world state to log entgries

ZephyrTFA committed:
world state has full timestamp, entry has readable timestamp

dragomagol commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,92 @@
-#define LOG_CATEGORY_NOT_FOUND "invalid_category"
-#define LOG_CATEGORY_TARGET_ZONE_SWITCH "target_zone_switch"
+// Misc categories
+#define LOG_CATEGORY_NOT_FOUND "invalid-category"
```
The lower groupings are alphabetized, would be nice if misc was as well

dragomagol commented on code/__HELPERS/text.dm:
```diff
@@ -1193,3 +1193,15 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
+		text2num(semver_regex.group[1]),
+		text2num(semver_regex.group[2]),
+		text2num(semver_regex.group[3]),
+		)
```
```suggestion
	)
```

dragomagol commented on code/modules/logging/categories/log_category_misc.dm:
```diff
@@ -0,0 +1,93 @@
+/datum/log_category/virus
```
I don't love that there's a misc category, especially one this big. The debug/development logs (runtime/qdel/error etc) can probably split out and the rest alphabetized.

ZephyrTFA committed:
Update code/__HELPERS/text.dm

Co-authored-by: tattle <66640614+dragomagol@users.noreply.github.com>

ZephyrTFA committed:
Update log_category_uplink.dm

Mothblocks commented:
Can you give an example output of a file that was JSON logged to a few times?

ZephyrTFA committed:
reorganize a bunch of shit
remove query category, we have sql
organize a bunch of misc into dedicated debug
sort misc

Signed-off-by: GitHub <noreply@github.com>

ZephyrTFA committed:
Merge branch 'master' into jlog

ZephyrTFA commented:
Here is `game.json`
```
{"timestamp":"1678519523.5136042","round-id":"0","secret":0,"schema-list":{"game-vote":"0.0.0","game-emote":"0.0.0","game-radio-emote":"0.0.0","game-traitor":"0.0.0","game-say":"0.0.0","game-whisper":"0.0.0","game-ooc":"0.0.0","game-prayer":"0.0.0","game-access":"0.0.0","game":"0.0.0"},"category":"game"}
{"timestamp":"2023-03-11 02:25:25.798","category":"game","message":"Station Trait: Colored Assistants chosen for this round.","world-state":{"tick_usage":73.952,"tick_lag":0.5,"time":15,"timestamp":"1678519525.7987278"}}
{"timestamp":"2023-03-11 02:25:32.261","category":"game","message":"Hierophant's Arena loaded at 202,81,5","world-state":{"tick_usage":48.5352,"tick_lag":0.5,"time":45.5,"timestamp":"1678519532.2614367"}}
{"timestamp":"2023-03-11 02:25:32.281","category":"game","message":"Blood-Drunk Miner (Random) loaded at 49,102,5","world-state":{"tick_usage":88.243,"tick_lag":0.5,"time":45.5,"timestamp":"1678519532.2812903"}}
{"timestamp":"2023-03-11 02:25:32.284","category":"game","message":"Pulsating Tumor loaded at 102,94,5","world-state":{"tick_usage":95.5128,"tick_lag":0.5,"time":45.5,"timestamp":"1678519532.284925"}}
{"timestamp":"2023-03-11 02:25:32.293","category":"game","message":"Bileworm Nest loaded at 81,188,5","world-state":{"tick_usage":112.828,"tick_lag":0.5,"time":45.5,"timestamp":"1678519532.2935822"}}
{"timestamp":"2023-03-11 02:25:32.298","category":"game","message":"Pulsating Tumor loaded at 179,41,5","world-state":{"tick_usage":122.474,"tick_lag":0.5,"time":45.5,"timestamp":"1678519532.2984068"}}
{"timestamp":"2023-03-11 02:25:32.316","category":"game","message":"Makeshift Shelter loaded at 217,191,5","world-state":{"tick_usage":157.751,"tick_lag":0.5,"time":45.5,"timestamp":"1678519532.3160443"}}
{"timestamp":"2023-03-11 02:25:32.377","category":"game","message":"Biodome Clown Planet loaded at 63,73,5","world-state":{"tick_usage":80.0788,"tick_lag":0.5,"time":46,"timestamp":"1678519532.377386"}}
{"timestamp":"2023-03-11 02:25:32.387","category":"game","message":"Ruin of Pride loaded at 212,26,5","world-state":{"tick_usage":101.256,"tick_lag":0.5,"time":46,"timestamp":"1678519532.3879752"}}
{"timestamp":"2023-03-11 02:25:32.404","category":"game","message":"UFO Crash loaded at 179,107,5","world-state":{"tick_usage":133.833,"tick_lag":0.5,"time":46,"timestamp":"1678519532.4042633"}}
{"timestamp":"2023-03-11 02:25:32.451","category":"game","message":"Biodome Winter loaded at 38,136,5","world-state":{"tick_usage":29.539,"tick_lag":0.5,"time":46.5,"timestamp":"1678519532.4519157"}}
{"timestamp":"2023-03-11 02:25:32.472","category":"game","message":"Ruin of Sloth loaded at 176,152,5","world-state":{"tick_usage":69.9586,"tick_lag":0.5,"time":46.5,"timestamp":"1678519532.4721236"}}
{"timestamp":"2023-03-11 02:25:32.478","category":"game","message":"Ripley loaded at 167,150,5","world-state":{"tick_usage":81.8056,"tick_lag":0.5,"time":46.5,"timestamp":"1678519532.4780476"}}
{"timestamp":"2023-03-11 02:25:32.492","category":"game","message":"Dark Wizard Altar loaded at 150,102,5","world-state":{"tick_usage":110.953,"tick_lag":0.5,"time":46.5,"timestamp":"1678519532.492621"}}
{"timestamp":"2023-03-11 02:25:32.683","category":"game","message":"A Giant Ball of Paper in Space loaded at 51,47,3","world-state":{"tick_usage":114.3,"tick_lag":0.5,"time":47,"timestamp":"1678519532.6836011"}}
{"timestamp":"2023-03-11 02:25:32.869","category":"game","message":"The Twin-Nexus Hotel loaded at 111,53,6","world-state":{"tick_usage":87.0904,"tick_lag":0.5,"time":48,"timestamp":"1678519532.8694687"}}
{"timestamp":"2023-03-11 02:25:32.880","category":"game","message":"Asteroid 1 loaded at 129,106,8","world-state":{"tick_usage":108.932,"tick_lag":0.5,"time":48,"timestamp":"1678519532.880389"}}
{"timestamp":"2023-03-11 02:25:32.898","category":"game","message":"Unnamed Turreted Outpost loaded at 173,210,8","world-state":{"tick_usage":144.47,"tick_lag":0.5,"time":48,"timestamp":"1678519532.898158"}}
{"timestamp":"2023-03-11 02:25:32.908","category":"game","message":"Authorship loaded at 181,219,9","world-state":{"tick_usage":165.24,"tick_lag":0.5,"time":48,"timestamp":"1678519532.9085426"}}
{"timestamp":"2023-03-11 02:25:32.998","category":"game","message":"Syndicate Ambush loaded at 137,168,6","world-state":{"tick_usage":139.632,"tick_lag":0.5,"time":48.5,"timestamp":"1678519532.998467"}}
{"timestamp":"2023-03-11 02:25:33.033","category":"game","message":"DK Excavator 453 loaded at 115,143,3","world-state":{"tick_usage":9.3704,"tick_lag":0.5,"time":49,"timestamp":"1678519533.033072"}}
{"timestamp":"2023-03-11 02:25:33.046","category":"game","message":"Asteroid 3 loaded at 66,40,8","world-state":{"tick_usage":35.3888,"tick_lag":0.5,"time":49,"timestamp":"1678519533.046081"}}
{"timestamp":"2023-03-11 02:25:33.060","category":"game","message":"Asteroid 6 loaded at 208,110,6","world-state":{"tick_usage":64.0912,"tick_lag":0.5,"time":49,"timestamp":"1678519533.0604322"}}
{"timestamp":"2023-03-11 02:25:33.102","category":"game","message":"Gondoland loaded at 146,68,7","world-state":{"tick_usage":147.55,"tick_lag":0.5,"time":49,"timestamp":"1678519533.102162"}}
{"timestamp":"2023-03-11 02:25:33.118","category":"game","message":"Cleric's Den loaded at 47,155,6","world-state":{"tick_usage":179.565,"tick_lag":0.5,"time":49,"timestamp":"1678519533.1181693"}}
{"timestamp":"2023-03-11 02:25:33.416","category":"game","message":"Ancient Space Station loaded at 27,101,9","world-state":{"tick_usage":138.989,"tick_lag":0.5,"time":50.5,"timestamp":"1678519533.416052"}}
{"timestamp":"2023-03-11 02:25:33.431","category":"game","message":"Empty Shell loaded at 82,70,9","world-state":{"tick_usage":170.767,"tick_lag":0.5,"time":50.5,"timestamp":"1678519533.4319408"}}
{"timestamp":"2023-03-11 02:25:33.460","category":"game","message":"Derelict 4 loaded at 80,102,8","world-state":{"tick_usage":29.2928,"tick_lag":0.5,"time":51,"timestamp":"1678519533.4609425"}}
{"timestamp":"2023-03-11 02:25:33.486","category":"game","message":"NT Medical Ship loaded at 57,124,7","world-state":{"tick_usage":80.3666,"tick_lag":0.5,"time":51,"timestamp":"1678519533.4864793"}}
{"timestamp":"2023-03-11 02:25:33.514","category":"game","message":"Clown Planet loaded at 150,161,8","world-state":{"tick_usage":136.115,"tick_lag":0.5,"time":51,"timestamp":"1678519533.5143538"}}
{"timestamp":"2023-03-11 02:25:41.135","category":"game","message":"*no key*/(Test Dummy) Unknown name changed from  to Vigistaezian ","world-state":{"tick_usage":2837.51,"tick_lag":0.5,"time":53.5,"timestamp":"1678519541.135913"}}
{"timestamp":"2023-03-11 02:25:42.498","category":"game","message":"*no key*/(monkey (246)) The monkey (246) name changed from monkey (246) to monkey (916) (Virology (62,82,2))","world-state":{"tick_usage":154.534,"tick_lag":0.5,"time":55,"timestamp":"1678519542.4986823"}}
{"timestamp":"2023-03-11 02:25:42.520","category":"game","message":"*no key*/(monkey (161)) The monkey (161) name changed from monkey (161) to monkey (694) (Virology (62,79,2))","world-state":{"tick_usage":198.494,"tick_lag":0.5,"time":55,"timestamp":"1678519542.5206623"}}
{"timestamp":"2023-03-11 02:25:42.545","category":"game","message":"*no key*/(monkey (26)) The monkey (26) name changed from monkey (26) to monkey (603) (Virology (63,83,2))","world-state":{"tick_usage":31.5108,"tick_lag":0.5,"time":55.5,"timestamp":"1678519542.5452683"}}
{"timestamp":"2023-03-11 02:25:42.565","category":"game","message":"*no key*/(monkey (986)) The monkey (986) name changed from monkey (986) to monkey (586) (Virology (63,78,2))","world-state":{"tick_usage":72.332,"tick_lag":0.5,"time":55.5,"timestamp":"1678519542.565679"}}
{"timestamp":"2023-03-11 02:25:42.586","category":"game","message":"*no key*/(monkey (123)) The monkey (123) name changed from monkey (123) to monkey (716) (Virology (64,80,2))","world-state":{"tick_usage":114.646,"tick_lag":0.5,"time":55.5,"timestamp":"1678519542.586836"}}
{"timestamp":"2023-03-11 02:25:42.817","category":"game","message":"*no key*/(monkey (770)) The monkey (770) name changed from monkey (770) to monkey (508) (Genetics Lab (110,80,2))","world-state":{"tick_usage":47.6002,"tick_lag":0.5,"time":56.5,"timestamp":"1678519542.8180914"}}
{"timestamp":"2023-03-11 02:25:42.840","category":"game","message":"*no key*/(monkey (758)) The monkey (758) name changed from monkey (758) to monkey (952) (Genetics Lab (110,79,2))","world-state":{"tick_usage":93.2298,"tick_lag":0.5,"time":56.5,"timestamp":"1678519542.8409052"}}
{"timestamp":"2023-03-11 02:25:42.861","category":"game","message":"*no key*/(monkey (498)) The monkey (498) name changed from monkey (498) to monkey (864) (Genetics Lab (111,79,2))","world-state":{"tick_usage":133.8,"tick_lag":0.5,"time":56.5,"timestamp":"1678519542.8611903"}}
{"timestamp":"2023-03-11 02:25:42.895","category":"game","message":"*no key*/(monkey (533)) The monkey (533) name changed from monkey (533) to Jimmy Rustles (Bar (133,127,2))","world-state":{"tick_usage":201.78,"tick_lag":0.5,"time":56.5,"timestamp":"1678519542.8951802"}}
{"timestamp":"2023-03-11 02:25:45.946","category":"game","message":"*no key*/(Julissa Sheets) Julissa Sheets name changed from  to Nameless Intern (CentCom Central Control (174,90,1))","world-state":{"tick_usage":188.335,"tick_lag":0.5,"time":68.5,"timestamp":"1678519545.9461954"}}
{"timestamp":"2023-03-11 02:25:56.761","category":"game","message":"*no key*/(Kerensa Jewell) Kerensa Jewell name changed from  to alien (Unexplored Location (185,112,5))","world-state":{"tick_usage":165.279,"tick_lag":0.5,"time":117.5,"timestamp":"1678519556.7612164"}}
{"timestamp":"2023-03-11 02:25:59.261","category":"game","message":"*no key*/(Homer Magor) Homer Magor name changed from Homer Magor to Greedo Shotfurst (Clown Planet (155,167,8))","world-state":{"tick_usage":208.611,"tick_lag":0.5,"time":129.5,"timestamp":"1678519559.2618675"}}
{"timestamp":"2023-03-11 02:25:59.319","category":"game","message":"*no key*/(Tristan Conrad) Tristan Conrad name changed from Tristan Conrad to Shoe Giving Gerald (Clown Planet (156,165,8))","world-state":{"tick_usage":102.972,"tick_lag":0.5,"time":130,"timestamp":"1678519559.3198013"}}
{"timestamp":"2023-03-11 02:25:59.363","category":"game","message":"*no key*/(Paige Harrold) Paige Harrold name changed from Paige Harrold to Jo Jo Bobo Bo (Clown Planet (163,163,8))","world-state":{"tick_usage":191.218,"tick_lag":0.5,"time":130,"timestamp":"1678519559.3639235"}}
{"timestamp":"2023-03-11 02:25:59.409","category":"game","message":"*no key*/(Luanne Sauter) Luanne Sauter name changed from Luanne Sauter to Slippy Joe (Clown Planet (168,166,8))","world-state":{"tick_usage":75.9128,"tick_lag":0.5,"time":130.5,"timestamp":"1678519559.4093153"}}
{"timestamp":"2023-03-11 02:26:12.877","category":"game-emote","message":"*no key*/(Audrina Huey) seizes up and falls limp, her eyes dead and lifeless... (Morgue (104,80,2))","world-state":{"tick_usage":1743.11,"tick_lag":0.5,"time":172,"timestamp":"1678519572.877606"}}
{"timestamp":"2023-03-11 02:26:12.934","category":"game","message":"*no key*/(Gunner Bashline) Gunner Bashline name changed from Gunner Bashline to Isochorista Fraterna (Morgue (100,81,2))","world-state":{"tick_usage":1856.61,"tick_lag":0.5,"time":172,"timestamp":"1678519572.934353"}}
{"timestamp":"2023-03-11 02:26:12.938","category":"game-emote","message":"*no key*/(Isochorista Fraterna) seizes up and falls limp, his eyes dead and lifeless... (Morgue (100,81,2))","world-state":{"tick_usage":1865.08,"tick_lag":0.5,"time":172,"timestamp":"1678519572.938588"}}
{"timestamp":"2023-03-11 02:26:20.915","category":"game","message":"Resetting persistent /datum/chemical_reaction/randomized/secret_sauce random recipe.","world-state":{"tick_usage":186.217,"tick_lag":0.5,"time":200,"timestamp":"1678519580.915802"}}
{"timestamp":"2023-03-11 02:26:20.917","category":"game","message":"Resetting persistent /datum/chemical_reaction/randomized/metalgen random recipe.","world-state":{"tick_usage":189.228,"tick_lag":0.5,"time":200,"timestamp":"1678519580.9173064"}}
{"timestamp":"2023-03-11 02:26:37.336","category":"game","message":"Backup Shuttle loaded at 17,17,12","world-state":{"tick_usage":96.1666,"tick_lag":0.5,"time":247,"timestamp":"1678519597.336841"}}
{"timestamp":"2023-03-11 02:26:37.458","category":"game","message":"fancy transport ferry loaded at 27,17,12","world-state":{"tick_usage":139.969,"tick_lag":0.5,"time":247.5,"timestamp":"1678519597.4584713"}}
{"timestamp":"2023-03-11 02:26:37.910","category":"game","message":"Meta Station Emergency Shuttle loaded at 42,17,12","world-state":{"tick_usage":203.741,"tick_lag":0.5,"time":249.5,"timestamp":"1678519597.910579"}}
{"timestamp":"2023-03-11 02:26:38.320","category":"game","message":"cargo ferry (Box) loaded at 69,17,12","world-state":{"tick_usage":200.348,"tick_lag":0.5,"time":251,"timestamp":"1678519598.320805"}}
{"timestamp":"2023-03-11 02:26:38.423","category":"game","message":"escape pod (Default) loaded at 78,17,12","world-state":{"tick_usage":185.187,"tick_lag":0.5,"time":251.5,"timestamp":"1678519598.423976"}}
{"timestamp":"2023-03-11 02:26:38.500","category":"game","message":"auxilliary base (Default) loaded at 83,17,12","world-state":{"tick_usage":136.821,"tick_lag":0.5,"time":252,"timestamp":"1678519598.5005364"}}
{"timestamp":"2023-03-11 02:26:38.662","category":"game","message":"arrival shuttle (Box) loaded at 94,17,12","world-state":{"tick_usage":57.6702,"tick_lag":0.5,"time":253,"timestamp":"1678519598.6628401"}}
{"timestamp":"2023-03-11 02:26:39.006","category":"game","message":"lavaland shuttle (Meta) loaded at 111,17,12","world-state":{"tick_usage":130.666,"tick_lag":0.5,"time":253.5,"timestamp":"1678519599.00608"}}
{"timestamp":"2023-03-11 02:26:39.101","category":"game","message":"mining shuttle (Delta) loaded at 118,17,12","world-state":{"tick_usage":119.328,"tick_lag":0.5,"time":254,"timestamp":"1678519599.1014662"}}
{"timestamp":"2023-03-11 02:26:39.201","category":"game","message":"labour shuttle (Delta) loaded at 125,17,12","world-state":{"tick_usage":117.326,"tick_lag":0.5,"time":254.5,"timestamp":"1678519599.2012665"}}
{"timestamp":"2023-03-11 02:26:39.273","category":"game","message":"escape pod (Default) loaded at 132,17,12","world-state":{"tick_usage":62.055,"tick_lag":0.5,"time":255,"timestamp":"1678519599.2732668"}}
{"timestamp":"2023-03-11 02:26:39.308","category":"game","message":"escape pod (Default) loaded at 137,17,12","world-state":{"tick_usage":132.968,"tick_lag":0.5,"time":255,"timestamp":"1678519599.3087237"}}
{"timestamp":"2023-03-11 02:26:39.353","category":"game","message":"escape pod (Default) loaded at 142,17,12","world-state":{"tick_usage":22.9608,"tick_lag":0.5,"time":255.5,"timestamp":"1678519599.3538947"}}
{"timestamp":"2023-03-11 02:26:39.494","category":"game","message":"Syndicate Fighter loaded at 147,17,12","world-state":{"tick_usage":15.0094,"tick_lag":0.5,"time":256,"timestamp":"1678519599.4946606"}}
{"timestamp":"2023-03-11 02:26:39.975","category":"game","message":"Small Freighter loaded at 154,17,12","world-state":{"tick_usage":46.3638,"tick_lag":0.5,"time":258,"timestamp":"1678519599.9759731"}}
{"timestamp":"2023-03-11 02:26:40.611","category":"game","message":"Pirate Cutter loaded at 177,17,12","world-state":{"tick_usage":225.461,"tick_lag":0.5,"time":260,"timestamp":"1678519600.611246"}}
{"timestamp":"2023-03-11 02:26:41.029","category":"game","message":"Syndicate Dropship loaded at 201,17,12","world-state":{"tick_usage":142.09,"tick_lag":0.5,"time":262,"timestamp":"1678519601.029981"}}
```

ZephyrTFA committed:
schema is in the log entry itself, header just contains subcats

ZephyrTFA commented:
changes per mothblocks request, (schema is now in the log entry and not the header)
```
{"timestamp":"1678521908.702898","round-id":"0","secret":0,"category-list":{"game-vote":null,"game-emote":null,"game-radio-emote":null,"game-traitor":null,"game-say":null,"game-whisper":null,"game-ooc":null,"game-prayer":null,"game-access":null,"game":"0.0.0"},"category":"game"}
{"timestamp":"2023-03-11 03:05:13.455","category":"game","message":"Hierophant's Arena loaded at 39,146,5","schema-version":"0.0.0","world-state":{"tick_usage":96.291,"tick_lag":0.5,"time":27.5,"timestamp":"1678521913.4551947"}}
{"timestamp":"2023-03-11 03:05:13.471","category":"game","message":"Blood-Drunk Miner (Random) loaded at 141,185,5","schema-version":"0.0.0","world-state":{"tick_usage":129.651,"tick_lag":0.5,"time":27.5,"timestamp":"1678521913.471875"}}
{"timestamp":"2023-03-11 03:05:13.475","category":"game","message":"Pulsating Tumor loaded at 207,57,5","schema-version":"0.0.0","world-state":{"tick_usage":137.196,"tick_lag":0.5,"time":27.5,"timestamp":"1678521913.4756472"}}
{"timestamp":"2023-03-11 03:05:13.497","category":"game","message":"Seed Vault loaded at 76,111,5","schema-version":"0.0.0","world-state":{"tick_usage":180.112,"tick_lag":0.5,"time":27.5,"timestamp":"1678521913.4971058"}}
{"timestamp":"2023-03-11 03:05:13.506","category":"game","message":"Survival Capsule Ruins loaded at 136,154,5","schema-version":"0.0.0","world-state":{"tick_usage":2.1006,"tick_lag":0.5,"time":28,"timestamp":"1678521913.5067272"}}
{"timestamp":"2023-03-11 03:05:13.515","category":"game","message":"Ruin of Envy loaded at 49,65,5","schema-version":"0.0.0","world-state":{"tick_usage":19.2682,"tick_lag":0.5,"time":28,"timestamp":"1678521913.515311"}}
{"timestamp":"2023-03-11 03:05:13.530","category":"game","message":"Ruin of Greed loaded at 175,77,5","schema-version":"0.0.0","world-state":{"tick_usage":49.2288,"tick_lag":0.5,"time":28,"timestamp":"1678521913.5302913"}}
{"timestamp":"2023-03-11 03:05:13.562","category":"game","message":"Ash Walker Nest loaded at 194,181,5","schema-version":"0.0.0","world-state":{"tick_usage":113.138,"tick_lag":0.5,"time":28,"timestamp":"1678521913.5622456"}}
{"timestamp":"2023-03-11 03:05:13.656","category":"game","message":"Strange Ship loaded at 116,158,8","schema-version":"0.0.0","world-state":{"tick_usage":15.8092,"tick_lag":0.5,"time":28.5,"timestamp":"1678521913.6562505"}}
{"timestamp":"2023-03-11 03:05:13.683","category":"game","message":"Heck Brewery loaded at 178,55,6","schema-version":"0.0.0","world-state":{"tick_usage":71.0086,"tick_lag":0.5,"time":28.5,"timestamp":"1678521913.6838505"}}
{"timestamp":"2023-03-11 03:05:13.698","category":"game","message":"Unnamed Turreted Outpost loaded at 53,190,6","schema-version":"0.0.0","world-state":{"tick_usage":99.452,"tick_lag":0.5,"time":28.5,"timestamp":"1678521913.6980722"}}
{"timestamp":"2023-03-11 03:05:13.748","category":"game","message":"Syndicate Forgotten Ship loaded at 45,39,7","schema-version":"0.0.0","world-state":{"tick_usage":199.406,"tick_lag":0.5,"time":28.5,"timestamp":"1678521913.7480488"}}
{"timestamp":"2023-03-11 03:05:13.765","category":"game","message":"Derelict 7 loaded at 173,100,8","schema-version":"0.0.0","world-state":{"tick_usage":24.5626,"tick_lag":0.5,"time":29,"timestamp":"1678521913.7651591"}}
{"timestamp":"2023-03-11 03:05:13.771","category":"game","message":"Asteroid 1 loaded at 43,57,9","schema-version":"0.0.0","world-state":{"tick_usage":36.3742,"tick_lag":0.5,"time":29,"timestamp":"1678521913.7710648"}}
{"timestamp":"2023-03-11 03:05:13.778","category":"game","message":"Authorship loaded at 133,80,9","schema-version":"0.0.0","world-state":{"tick_usage":51.298,"tick_lag":0.5,"time":29,"timestamp":"1678521913.7785265"}}
{"timestamp":"2023-03-11 03:05:13.786","category":"game","message":"CF Corsair loaded at 223,48,10","schema-version":"0.0.0","world-state":{"tick_usage":68.1164,"tick_lag":0.5,"time":29,"timestamp":"1678521913.786936"}}
{"timestamp":"2023-03-11 03:05:13.799","category":"game","message":"Cleric's Den loaded at 134,136,6","schema-version":"0.0.0","world-state":{"tick_usage":93.557,"tick_lag":0.5,"time":29,"timestamp":"1678521913.7996562"}}
{"timestamp":"2023-03-11 03:05:13.819","category":"game","message":"Biological Storage Facility loaded at 114,192,6","schema-version":"0.0.0","world-state":{"tick_usage":134.262,"tick_lag":0.5,"time":29,"timestamp":"1678521913.8200085"}}
{"timestamp":"2023-03-11 03:05:13.847","category":"game","message":"Derelict 5 loaded at 128,106,9","schema-version":"0.0.0","world-state":{"tick_usage":189.28,"tick_lag":0.5,"time":29,"timestamp":"1678521913.847518"}}
{"timestamp":"2023-03-11 03:05:13.855","category":"game","message":"Detached Teleporter loaded at 209,44,9","schema-version":"0.0.0","world-state":{"tick_usage":7.2734,"tick_lag":0.5,"time":29.5,"timestamp":"1678521913.8553586"}}
{"timestamp":"2023-03-11 03:05:13.882","category":"game","message":"Syndicate Listening Station loaded at 179,54,7","schema-version":"0.0.0","world-state":{"tick_usage":61.6146,"tick_lag":0.5,"time":29.5,"timestamp":"1678521913.8825295"}}
{"timestamp":"2023-03-11 03:05:13.891","category":"game","message":"Asteroid 4 loaded at 78,98,4","schema-version":"0.0.0","world-state":{"tick_usage":79.5454,"tick_lag":0.5,"time":29.5,"timestamp":"1678521913.8914948"}}
{"timestamp":"2023-03-11 03:05:13.899","category":"game","message":"Dinner for Two loaded at 25,33,7","schema-version":"0.0.0","world-state":{"tick_usage":94.762,"tick_lag":0.5,"time":29.5,"timestamp":"1678521913.8991032"}}
{"timestamp":"2023-03-11 03:05:13.914","category":"game","message":"Waylaid Buses loaded at 145,184,6","schema-version":"0.0.0","world-state":{"tick_usage":126.145,"tick_lag":0.5,"time":29.5,"timestamp":"1678521913.9147944"}}
{"timestamp":"2023-03-11 03:05:19.495","category":"game","message":"*no key*/(Test Dummy) Unknown name changed from  to Vigistaezian ","schema-version":"0.0.0","world-state":{"tick_usage":2055.1,"tick_lag":0.5,"time":32,"timestamp":"1678521919.4956806"}}
{"timestamp":"2023-03-11 03:05:20.509","category":"game","message":"*no key*/(monkey (638)) The monkey (638) name changed from monkey (638) to monkey (440) (Virology (62,82,2))","schema-version":"0.0.0","world-state":{"tick_usage":41.009,"tick_lag":0.5,"time":33.5,"timestamp":"1678521920.5094328"}}
{"timestamp":"2023-03-11 03:05:20.525","category":"game","message":"*no key*/(monkey (590)) The monkey (590) name changed from monkey (590) to monkey (985) (Virology (62,79,2))","schema-version":"0.0.0","world-state":{"tick_usage":74.0198,"tick_lag":0.5,"time":33.5,"timestamp":"1678521920.5259376"}}
{"timestamp":"2023-03-11 03:05:20.542","category":"game","message":"*no key*/(monkey (452)) The monkey (452) name changed from monkey (452) to monkey (723) (Virology (63,83,2))","schema-version":"0.0.0","world-state":{"tick_usage":106.386,"tick_lag":0.5,"time":33.5,"timestamp":"1678521920.542121"}}
{"timestamp":"2023-03-11 03:05:20.557","category":"game","message":"*no key*/(monkey (639)) The monkey (639) name changed from monkey (639) to monkey (910) (Virology (63,78,2))","schema-version":"0.0.0","world-state":{"tick_usage":137.918,"tick_lag":0.5,"time":33.5,"timestamp":"1678521920.5578866"}}
{"timestamp":"2023-03-11 03:05:20.573","category":"game","message":"*no key*/(monkey (494)) The monkey (494) name changed from monkey (494) to monkey (813) (Virology (64,80,2))","schema-version":"0.0.0","world-state":{"tick_usage":169.638,"tick_lag":0.5,"time":33.5,"timestamp":"1678521920.5737467"}}
{"timestamp":"2023-03-11 03:05:20.745","category":"game","message":"*no key*/(monkey (65)) The monkey (65) name changed from monkey (65) to monkey (789) (Genetics Lab (110,80,2))","schema-version":"0.0.0","world-state":{"tick_usage":64.175,"tick_lag":0.5,"time":34.5,"timestamp":"1678521920.7451086"}}
{"timestamp":"2023-03-11 03:05:20.760","category":"game","message":"*no key*/(monkey (362)) The monkey (362) name changed from monkey (362) to monkey (783) (Genetics Lab (110,79,2))","schema-version":"0.0.0","world-state":{"tick_usage":95.7762,"tick_lag":0.5,"time":34.5,"timestamp":"1678521920.7609093"}}
{"timestamp":"2023-03-11 03:05:20.777","category":"game","message":"*no key*/(monkey (318)) The monkey (318) name changed from monkey (318) to monkey (507) (Genetics Lab (111,79,2))","schema-version":"0.0.0","world-state":{"tick_usage":128.712,"tick_lag":0.5,"time":34.5,"timestamp":"1678521920.7773774"}}
{"timestamp":"2023-03-11 03:05:20.800","category":"game","message":"*no key*/(monkey (125)) The monkey (125) name changed from monkey (125) to Jimmy Rustles (Bar (133,127,2))","schema-version":"0.0.0","world-state":{"tick_usage":175.933,"tick_lag":0.5,"time":34.5,"timestamp":"1678521920.800988"}}
{"timestamp":"2023-03-11 03:05:22.982","category":"game","message":"*no key*/(Georgia Levett) Georgia Levett name changed from  to Nameless Intern (CentCom Central Control (174,90,1))","schema-version":"0.0.0","world-state":{"tick_usage":102.731,"tick_lag":0.5,"time":43.5,"timestamp":"1678521922.982343"}}
{"timestamp":"2023-03-11 03:05:31.956","category":"game","message":"*no key*/(Rihanna Ulery) Rihanna Ulery name changed from Rihanna Ulery to Pogo (Unexplored Location (89,101,4))","schema-version":"0.0.0","world-state":{"tick_usage":228.275,"tick_lag":0.5,"time":84.5,"timestamp":"1678521931.9560742"}}
{"timestamp":"2023-03-11 03:05:40.037","category":"game-emote","message":"*no key*/(Solomon Kemble) seizes up and falls limp, his eyes dead and lifeless... (Morgue (102,78,2))","schema-version":"0.0.0","world-state":{"tick_usage":1322.7,"tick_lag":0.5,"time":116,"timestamp":"1678521940.0380044"}}
{"timestamp":"2023-03-11 03:05:40.080","category":"game","message":"*no key*/(Janine Aggley) Janine Aggley name changed from Janine Aggley to Emilia Lowstetter (Morgue (104,81,2))","schema-version":"0.0.0","world-state":{"tick_usage":1408.6,"tick_lag":0.5,"time":116,"timestamp":"1678521940.0809526"}}
{"timestamp":"2023-03-11 03:05:40.083","category":"game-emote","message":"*no key*/(Emilia Lowstetter) seizes up and falls limp, her eyes dead and lifeless... (Morgue (104,81,2))","schema-version":"0.0.0","world-state":{"tick_usage":1414.5,"tick_lag":0.5,"time":116,"timestamp":"1678521940.0839014"}}
{"timestamp":"2023-03-11 03:05:45.168","category":"game","message":"Resetting persistent /datum/chemical_reaction/randomized/secret_sauce random recipe.","schema-version":"0.0.0","world-state":{"tick_usage":88.7342,"tick_lag":0.5,"time":133.5,"timestamp":"1678521945.1682415"}}
{"timestamp":"2023-03-11 03:05:45.169","category":"game","message":"Resetting persistent /datum/chemical_reaction/randomized/metalgen random recipe.","schema-version":"0.0.0","world-state":{"tick_usage":91.2924,"tick_lag":0.5,"time":133.5,"timestamp":"1678521945.1695206"}}
{"timestamp":"2023-03-11 03:05:57.072","category":"game","message":"Backup Shuttle loaded at 17,17,12","schema-version":"0.0.0","world-state":{"tick_usage":117.793,"tick_lag":0.5,"time":161.5,"timestamp":"1678521957.072229"}}
{"timestamp":"2023-03-11 03:05:57.145","category":"game","message":"fancy transport ferry loaded at 27,17,12","schema-version":"0.0.0","world-state":{"tick_usage":44.24,"tick_lag":0.5,"time":162,"timestamp":"1678521957.145089"}}
{"timestamp":"2023-03-11 03:05:57.433","category":"game","message":"Meta Station Emergency Shuttle loaded at 42,17,12","schema-version":"0.0.0","world-state":{"tick_usage":165.859,"tick_lag":0.5,"time":163,"timestamp":"1678521957.4334824"}}
{"timestamp":"2023-03-11 03:05:57.659","category":"game","message":"cargo ferry (Box) loaded at 69,17,12","schema-version":"0.0.0","world-state":{"tick_usage":39.004,"tick_lag":0.5,"time":164,"timestamp":"1678521957.6592822"}}
{"timestamp":"2023-03-11 03:05:57.718","category":"game","message":"escape pod (Default) loaded at 78,17,12","schema-version":"0.0.0","world-state":{"tick_usage":157.808,"tick_lag":0.5,"time":164,"timestamp":"1678521957.718684"}}
{"timestamp":"2023-03-11 03:05:57.772","category":"game","message":"auxilliary base (Default) loaded at 83,17,12","schema-version":"0.0.0","world-state":{"tick_usage":45.9862,"tick_lag":0.5,"time":164.5,"timestamp":"1678521957.7720625"}}
{"timestamp":"2023-03-11 03:05:57.876","category":"game","message":"arrival shuttle (Box) loaded at 94,17,12","schema-version":"0.0.0","world-state":{"tick_usage":56.4946,"tick_lag":0.5,"time":165,"timestamp":"1678521957.877036"}}
{"timestamp":"2023-03-11 03:05:57.955","category":"game","message":"lavaland shuttle (Meta) loaded at 111,17,12","schema-version":"0.0.0","world-state":{"tick_usage":13.5956,"tick_lag":0.5,"time":165.5,"timestamp":"1678521957.9553792"}}
{"timestamp":"2023-03-11 03:05:58.012","category":"game","message":"mining shuttle (Delta) loaded at 118,17,12","schema-version":"0.0.0","world-state":{"tick_usage":127.447,"tick_lag":0.5,"time":165.5,"timestamp":"1678521958.0123045"}}
{"timestamp":"2023-03-11 03:05:58.085","category":"game","message":"labour shuttle (Delta) loaded at 125,17,12","schema-version":"0.0.0","world-state":{"tick_usage":59.2068,"tick_lag":0.5,"time":166,"timestamp":"1678521958.0850813"}}
{"timestamp":"2023-03-11 03:05:58.130","category":"game","message":"escape pod (Default) loaded at 132,17,12","schema-version":"0.0.0","world-state":{"tick_usage":149.302,"tick_lag":0.5,"time":166,"timestamp":"1678521958.1301284"}}
{"timestamp":"2023-03-11 03:05:58.157","category":"game","message":"escape pod (Default) loaded at 137,17,12","schema-version":"0.0.0","world-state":{"tick_usage":4.6814,"tick_lag":0.5,"time":166.5,"timestamp":"1678521958.1577363"}}
{"timestamp":"2023-03-11 03:05:58.183","category":"game","message":"escape pod (Default) loaded at 142,17,12","schema-version":"0.0.0","world-state":{"tick_usage":56.0306,"tick_lag":0.5,"time":166.5,"timestamp":"1678521958.183411"}}
```

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into jlog

Mothblocks reviewed (approved):
Looks fine but I'm not permitted to merge admin stuff since I'm not in the Discord

ZephyrTFA committed:
forgot to change this one

optimumtact commented:
@MrStonedOne do you have any more comments?

MrStonedOne commented:
.... maybe. let me take the time to do another thorough lookthru

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,97 @@
+#define LOG_CATEGORY_ADMIN_DSAY "admin-dsay"
+
+// Admin private categories
+#define LOG_CATEGORY_ADMIN_PRIVATE "admin-private"
+#define LOG_CATEGORY_ADMIN_PRIVATE_ASAY "admin-asay"
```
Could you please Rename admin-private to adminprivate so public logs can easy parse `-` as a analog for `LOGTYPE: LOGSUBTYPE:` and i can reuse the same switch tree for both file types when i add support to public logs.

asay should also be renamed to `"adminprivate-asay"` to keep it inline with the above as well.

basically

ZephyrTFA replied:
gotcha

ZephyrTFA committed:
Update logging.dm

MrStonedOne commented on code/__HELPERS/logging/admin.dm:
```diff
@@ -3,31 +3,26 @@
 /proc/log_admin_circuit(text)
 	GLOB.admin_log.Add(text)
-	if(CONFIG_GET(flag/log_admin))
-		WRITE_LOG(GLOB.world_game_log, "ADMIN: CIRCUIT: [text]")
+	logger.Log(LOG_CATEGORY_ADMIN_CIRCUIT, text)
 
 /// General logging for admin actions
 /proc/log_admin_private(text)
 	GLOB.admin_log.Add(text)
-	if (CONFIG_GET(flag/log_admin))
-		WRITE_LOG(GLOB.world_game_log, "ADMINPRIVATE: [text]")
+	logger.Log(LOG_CATEGORY_ADMIN_PRIVATE, text)
 
 /// Logging for AdminSay (ASAY) messages
 /proc/log_adminsay(text)
 	GLOB.admin_log.Add(text)
-	if (CONFIG_GET(flag/log_adminchat))
-		WRITE_LOG(GLOB.world_game_log, "ADMINPRIVATE: ASAY: [text]")
+	logger.Log(LOG_CATEGORY_ADMIN_PRIVATE_ASAY, text)
 
 /// Logging for DeachatSay (DSAY) messages
 /proc/log_dsay(text)
-	if (CONFIG_GET(flag/log_adminchat))
-		WRITE_LOG(GLOB.world_game_log, "ADMIN: DSAY: [text]")
+	logger.Log(LOG_CATEGORY_ADMIN_DSAY, text)
 
```
for now can we please keep logs that go to the game.log file going to game.json logfile.

at least until full tooling and viewing merged files is properly set up.

an alternative could include sending it to both.

MrStonedOne commented on code/__HELPERS/logging/debug.dm:
```diff
@@ -38,30 +36,26 @@
 /* Log to the logfile only. */
 /proc/log_runtime(text)
-	WRITE_LOG(GLOB.world_runtime_log, text)
+	logger.Log(LOG_CATEGORY_DEBUG_RUNTIME, text)
```
lets leave runtime logs off, or maybe just do both for this one: runtime.log and runtime.json. Its gonna take me way too long to add support for parsing new runtimes given i have to edit both php and c++ code to do so and the c++ code is a single c file with no deps so i'd rather not have to figure out how to add json support.

we have dd.log as well, but thats unsanitized and includes things that can't be trusted to coders/players in public logs and also doesn't have the same timestamp format `rc.cpp` expects

MrStonedOne commented on code/modules/logging/categories/log_category_misc.dm:
```diff
@@ -0,0 +1,58 @@
+	category = LOG_CATEGORY_SIGNAL
+
+/datum/log_category/dynamic
+	category = LOG_CATEGORY_DYNAMIC
```
dynamic.json already exists as a log file.

this will conflict.

ZephyrTFA committed:
add backwards compat mirrors

ZephyrTFA committed:
Merge branch 'jlog' of github.com:ZephyrTFA/tgstation into jlog

ZephyrTFA committed:
add world runtime log back
and have inround dynamic logging sent to dynamic_gamemode

ZephyrTFA committed:
compat logging instead

ZephyrTFA committed:
touchups
move logging config to it's own config

Mothblocks commented on code/game/gamemodes/dynamic/dynamic_logging.dm:
```diff
@@ -81,7 +81,7 @@
 	serialized["snapshots"] = serialized_snapshots
 
-	rustg_file_write(json_encode(serialized), "[GLOB.log_directory]/dynamic.json")
+	rustg_file_write(json_encode(serialized), "[GLOB.log_directory]/dynamic_gamemode.json")
```
Why are we changing the existing one instead of the new one? This is not right

Mothblocks replied:
Especially since this new name is completely ambiguous

Mothblocks commented on code/controllers/master.dm:
```diff
@@ -14,6 +14,9 @@
 GLOBAL_REAL(Debugger, /datum/debugger) = new
 //This is the ABSOLUTE ONLY THING that should init globally like this
 //2019 update: the failsafe,config and Global controllers also do it
+//2023 update: the logger also does it
```
These historical comments are dumb, change the whole thing to say "Nothing but these should be initting globally like this"

I feel like I commented this before. Did you resolve it or did I not finish it?

ZephyrTFA replied:
Don't remember seeing a comment on that no

ZephyrTFA committed:
Update master.dm

ZephyrTFA committed:
handle semver collection better using a macro
assemble the semver list into the final entry as data

ZephyrTFA committed:
change new files to .log.json and revert changing existing

ZephyrTFA committed:
whoops

ZephyrTFA committed:
guh

Mothblocks commented on code/__HELPERS/logging/_logging.dm:
```diff
@@ -10,6 +10,31 @@
+#define CHECK_SERIALIZATION_SEMVER(wanted, actual) (__check_serialization_semver(wanted, actual))
+
+/// Checks if the actual semver is equal or later than the wanted semver
+/// Must be passed as TEXT
```
If you don't want people using this, point them to the macro

ZephyrTFA replied:
```suggestion
/// Must be passed as TEXT; you're probably looking for CHECK_SERIALIZATION_SEMVER, look right above
```

Mothblocks commented on code/__HELPERS/logging/_logging.dm:
```diff
@@ -10,6 +10,31 @@
 	msg = "## NOTICE: [msg]"
 	log_world(msg)
 
+#define SET_SERIALIZATION_SEMVER(semver_list, semver) semver_list[type] = semver
+#define CHECK_SERIALIZATION_SEMVER(wanted, actual) (__check_serialization_semver(wanted, actual))
```
These need doc comments

GoldenAlpharex replied:
You didn't document these two defines still though.

ZephyrTFA committed:
Update code/__HELPERS/logging/_logging.dm

TheVekter commented:
Can we avoid actually merging this until the log viewer is complete? This is actively making it harder to look through logs on Basil and I'm not looking forward to dealing with reading these until the tool they're meant for is done.

Mothblocks commented:
@TheVekter What's it doing to the log viewer? This change should've been completely internal, what do they look like?

Mothblocks commented:
This is meant to aide tools, but the in game tools should always be perfectly usable and you should never need external stuff to admin

MrStonedOne commented:
> @TheVekter What's it doing to the log viewer? This change should've been completely internal, what do they look like?

Most people view logs in notepad, the browser via raw-log or parsed-logs, or in the `view current logs` verb which just shows files inside of a `<pre>` tag,

They do not use the individual logs panel you are thinking of for most admin tasks. You should know this.

Adding all of the json bullshit does make shit less readable, yes.

> This is meant to aide tools, but the in game tools should always be perfectly usable and you should never need external stuff to admin

see (from pr op)

> I am also working on a log viewer frontend for admin usage however that will take a little bit longer to finish up.

MrStonedOne commented:
> what do they look like?

https://tgstation13.org/raw-logs/basil/data/logs/2023/03/14/round-201835/game.json

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
+#define LOG_ENTRY_SEMVER_STORE "semver-store"
 
 // Log json keys
 #define LOG_JSON_CATEGORY "category"
```
```suggestion
#define LOG_JSON_CATEGORY "cat"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
 
 // Log data keys
+#define LOG_ENTRY_CATEGORY "category"
 #define LOG_ENTRY_MESSAGE "message"
```
```suggestion
#define LOG_ENTRY_MESSAGE "msg"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
+#define LOG_HEADER_SECRET "secret"
 
 // Log data keys
+#define LOG_ENTRY_CATEGORY "category"
```
```suggestion
#define LOG_ENTRY_CATEGORY "cat"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
 // Log data keys
+#define LOG_ENTRY_CATEGORY "category"
 #define LOG_ENTRY_MESSAGE "message"
 #define LOG_ENTRY_TIMESTAMP "timestamp"
```
```suggestion
#define LOG_ENTRY_TIMESTAMP "ts"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
 // Log header keys
 #define LOG_HEADER_CATEGORY "category"
+#define LOG_HEADER_CATEGORY_LIST "category-list"
 #define LOG_HEADER_INIT_TIMESTAMP "timestamp"
```
```suggestion
#define LOG_HEADER_INIT_TIMESTAMP "ts"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
 
 // Log header keys
 #define LOG_HEADER_CATEGORY "category"
+#define LOG_HEADER_CATEGORY_LIST "category-list"
```
```suggestion
#define LOG_HEADER_CATEGORY_LIST "cat-list"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
 
 // Log header keys
 #define LOG_HEADER_CATEGORY "category"
```
```suggestion
#define LOG_HEADER_CATEGORY "cat"
```

MrStonedOne commented on code/__DEFINES/logging.dm:
```diff
@@ -56,22 +56,101 @@
+#define SCHEMA_VERSION "schema-version"
+
+// Default log schema version
+#define LOG_CATEGORY_SCHEMA_VERSION_NOT_SET "0.0.0"
```
```suggestion
#define LOG_CATEGORY_SCHEMA_VERSION_NOT_SET "0.0.1"
```

MrsTonedOne reviewed (commented):
Readability idea: (see suggested changes)

It might be a bit excessive to do this, i don't know, but it could help with readability of the raw log files.

Mothblocks commented:
> They do not use the individual logs panel you are thinking of for most admin tasks. You should know this.

My shock was that we weren't giving them plain text logs in the actual display text

Mothblocks commented:
>  I am also working on a log viewer frontend for admin usage however that will take a little bit longer to finish up.


Okay, this makes everything make way more sense, thank you

optimumtact commented:
I think there has been a miscommunication @Mothblocks my original desgn document intended to keep the text logs around for at least a whiel as an output format
```
every logging_entry datum must implement a to_text method, that writes a human readable form of the log entry, SSlog by default will always write to a relevant logging file this textual form, with a config flag available to turn this off if you only wish to present logs to end users via a webservice or other consuming party.
```

I think you took me complaining about having "two logging systems" on an earlier PR as meaning we can't have two output formats, which isn't what I meant.

I wanted to maintain the existing textual logs for at least a while because there are already significant third party tools admins use to parse logs that need time to update to a json api.

ZephyrTFA committed:
human readable

ZephyrTFA committed:
Apply suggestions from code review

Co-authored-by: Kyle Spier-Swenson <kyleshome@gmail.com>

ZephyrTFA committed:
whoops

ZephyrTFA committed:
Merge branch 'jlog' of github.com:ZephyrTFA/tgstation into jlog

ZephyrTFA committed:
better json encoding

Co-authored-by: Kyle Spier-Swenson <kyleshome@gmail.com>

ZephyrTFA committed:
add data param to logging helpers for future usage

ZephyrTFA committed:
copy paste moment

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into jlog

optimumtact commented:
@MrStonedOne do you have any final issues with this?

MrStonedOne commented:
this can't even be considered until the byond version that fixes json memory leaks is released.

I want to test merge this again after that

Cyberboss reviewed (changes requested):
As long as it isn't necessary, can you remove `shutdown_logging()` entirely? It makes world end awkward.

ZephyrTFA commented:
> As long as it isn't necessary, can you remove `shutdown_logging()` entirely? It makes world end awkward.

still required for `GLOB.world_runtime_log` `GLOB.test_log` and `GLOB.harddel_log`

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into jlog

ZephyrTFA committed:
ah fuck

ZephyrTFA committed:
re-refactor log files

san7890 commented:
I pulled the TM on Round 204841 (the link to the logs should be in the automated github entry) because they were causing a great heap of overtime, per this image:

![image](https://user-images.githubusercontent.com/34697715/235391546-f82e88ea-b922-4066-8eb0-109361096799.png)

According to reports, the lag started after highlander was enabled? I've also heard scattered reports of lag on sybil all day today, but i am not 100% sure if it was being triggered by that as well.

ZephyrTFA committed:
removed list of entries because list ops

ZephyrTFA commented:
Identified the likely culprit

ZephyrTFA committed:
Update log_category_debug.dm

ZephyrTFA commented on code/modules/logging/categories/log_category_debug.dm:
```diff
@@ -0,0 +1,45 @@
+// This is not in the debug master category on purpose, do not add it
+/datum/log_category/debug_runtime
+	category = LOG_CATEGORY_DEBUG_RUNTIME
+
```
```suggestion
```

ZephyrTFA committed:
Update code/modules/logging/categories/log_category_debug.dm

ZephyrTFA committed:
use logging procs to prevent memory waste

ZephyrTFA committed:
sad

ZephyrTFA committed:
main rustg formatting on demand, add to runtime and fix category ident

GoldenAlpharex commented on code/__HELPERS/logging/admin.dm:
```diff
@@ -36,8 +39,7 @@
-	if (CONFIG_GET(flag/log_suspicious_login))
-		WRITE_LOG(GLOB.world_suspicious_login_log, "SUSPICIOUS_ACCESS: [text]")
+/proc/log_suspicious_login(text, list/data, access_log_mirror = TRUE)
+	logger.Log(LOG_CATEGORY_SUSPICIOUS_LOGIN, text)
```
This doesn't respect the config flag anymore, which makes either the proc's documentation inaccurate, or is a regression.

GoldenAlpharex commented on code/__HELPERS/logging/antagonists.dm:
```diff
@@ -1,29 +1,23 @@
-	if (CONFIG_GET(flag/log_traitor))
-		WRITE_LOG(GLOB.world_game_log, "TRAITOR: [text]")
+/proc/log_traitor(text, list/data)
+	logger.Log(LOG_CATEGORY_GAME_TRAITOR, text, data)
```
Are these configs now deprecated since they're always logged following your system?

GoldenAlpharex commented on code/__HELPERS/logging/mob.dm:
```diff
@@ -3,12 +3,11 @@
-	if (CONFIG_GET(flag/log_silicon))
-		WRITE_LOG(GLOB.world_silicon_log, "SILICON: [text]")
+/proc/log_silicon(text, list/data)
+	logger.Log(LOG_CATEGORY_SILICON, text, data)
```
Same deal here about config.

GoldenAlpharex commented on code/__HELPERS/logging/pda.dm:
```diff
@@ -1,16 +1,11 @@
-	if (CONFIG_GET(flag/log_pda))
-		WRITE_LOG(GLOB.world_pda_log, "PDA: [text]")
+/proc/log_pda(text, list/data)
+	logger.Log(LOG_CATEGORY_PDA, text, data)
```
Ditto.

GoldenAlpharex commented on code/datums/datum.dm:
```diff
@@ -183,13 +183,20 @@
+	.["tag"] = tag
 
-///Accepts a LIST from deserialize_datum. Should return src or another datum.
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
```
Should this version not be a define or something? I'm not 100% as to what the versioning here stands for, mind you.

ZephyrTFA replied:
versions are not going to be a define

GoldenAlpharex commented on code/datums/datum.dm:
```diff
@@ -241,11 +248,10 @@
 	var/datum/D = new typeofdatum
-	var/datum/returned = D.deserialize_list(jsonlist, options)
-	if(!isdatum(returned))
+	if(!D.deserialize_list(jsonlist, options))
 		qdel(D)
 	else
-		return returned
+		return D
```
Single letter var :snapped:

ZephyrTFA replied:
existing code, not worth touching

GoldenAlpharex commented on code/datums/json_savefile.dm:
```diff
@@ -57,7 +57,9 @@ GENERAL_PROTECT_DATUM(/datum/json_savefile)
-/datum/json_savefile/serialize_list(list/options)
+/datum/json_savefile/serialize_list(list/options, list/semvers)
+	SHOULD_CALL_PARENT(FALSE)
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
```
Same here for the string version.

GoldenAlpharex commented on code/datums/martial/_martial.dm:
```diff
@@ -16,6 +16,16 @@
+	.["id"] = id
+	.["pacifist_style"] = pacifist_style
+
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
```
Ditto.

GoldenAlpharex commented on code/datums/memory/_memory.dm:
```diff
@@ -75,6 +75,22 @@
+	.["antagonist_name"] = antagonist_name
+	.["where"] = where
+
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
```
Ditto.

GoldenAlpharex commented on code/datums/mind/_mind.dm:
```diff
@@ -122,6 +122,25 @@
+	var/mob/enslaved_to = src.enslaved_to?.resolve()
+	.["enslaved_to"] = enslaved_to
+
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
```
Ditto.

GoldenAlpharex commented on code/modules/logging/log_category.dm:
```diff
@@ -1,32 +1,56 @@
+	/// Whether or not this log should not be publically visible
+	var/secret = FALSE
+
+	/// Whether the readable version of the runtime if formatted internally instead of by rustg
```
Did you mean `log` here, rather than `runtime`?

ZephyrTFA replied:
```suggestion
	/// Whether the readable version of the log message is formatted internally instead of by rustg
```

GoldenAlpharex commented on code/modules/logging/log_entry.dm:
```diff
@@ -0,0 +1,114 @@
+	/// Data of the log entry; optional.
+	var/list/data
+
+	/// Semver store of the log entry, used to store the schema of data entries
```
Could be nice to have an explanation somewhere of both what a Semver is, and what a Semver Store is. Because even I had to Google it, after I couldn't find an explanation on here.

GoldenAlpharex commented on code/modules/logging/log_entry.dm:
```diff
@@ -0,0 +1,114 @@
+/// A datum which contains log information.
+/datum/log_entry
+	/// Next id to assign to a log entry.
+	var/static/next_id = 0
```
So these IDs are unique within a round, but not across rounds?

And what happens if we overflow, is that possible?

ZephyrTFA replied:
if we manage to get over 256k+ log entries I'll be shocked

GoldenAlpharex commented on code/modules/logging/log_holder.dm:
```diff
@@ -61,10 +78,120 @@ GLOBAL_PROTECT(logger)
+	// Sort the sub categories
+	for(var/datum/log_category/master as anything in sub_categories)
+		if(!(master in category_tree))
+			stack_trace("log category [master] is an invalid master category as its a sub category")
```
```suggestion
			stack_trace("log category [master] is an invalid master category as it's a sub category")
```

GoldenAlpharex commented on code/modules/logging/log_holder.dm:
```diff
@@ -61,10 +78,120 @@ GLOBAL_PROTECT(logger)
+	if(human_readable_enabled)
+		rustg_file_write("\[[human_readable_timestamp()]\] Starting up round ID [round_id].\n - -------------------------\n", category_instance.get_output_file(null, "log"))
+
+/datum/log_holder/proc/unix_timestamp_string() // pending change to rust-g
+	return RUSTG_CALL(RUST_G, "unix_timestamp")()
```
Are those changes to rust-g still pending?

ZephyrTFA replied:
annoyingly yes

GoldenAlpharex commented on code/modules/mob/mob.dm:
```diff
@@ -109,6 +109,17 @@
+	.["ckey"] = ckey
+	.["key"] = key
+
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
```
Same deal here about the versioning not being in a define.

GoldenAlpharex commented on code/modules/photography/_pictures.dm:
```diff
@@ -69,9 +70,19 @@
 	.["blueprints"] = has_blueprints
 	.["logpath"] = logpath
 
+	SET_SERIALIZATION_SEMVER(semvers, "1.0.0")
+	return .
+
 /datum/picture/deserialize_list(list/input, list/options)
+	if((SCHEMA_VERSION in options) && (options[SCHEMA_VERSION] != "1.0.0"))
+		CRASH("Invalid schema version for datum/picture: [options[SCHEMA_VERSION]] (expected 1.0.0)")
```
Same deal here about the versioning not being in a define.

GoldenAlpharex reviewed (changes requested):
Just in case it gets burrowed - Would be nice to have a definition somewhere in the code of what a `semver` and a `semver store` are. It's rather confusing, and having to search on Google to figure out why it's called that is probably not the best idea in the world.

ZephyrTFA committed:
Update code/modules/logging/log_holder.dm

Co-authored-by: GoldenAlpharex <58045821+GoldenAlpharex@users.noreply.github.com>

ZephyrTFA committed:
Update code/modules/logging/log_category.dm
